colossalai/booster/plugin/moe_hybrid_parallel_plugin.py

```diff
@@ -220,70 +204,89 @@ def __init__(
         self.enable_flash_attention = enable_flash_attention
         self.enable_jit_fused = enable_jit_fused
         self.enable_sequence_parallelism = enable_sequence_parallelism
-        self.checkpoint_io = checkpoint_io
-
-        logger = get_dist_logger()
-
-        # NOTE: Two process meshes: global dp for non-moe param; dp + ep for moe param
-        # See https://hpc-ai.com/blog/enhanced-moe-parallelism-open-source-moe-model-training-can-be-9-times-more-efficient
-        # we change pg mesh to (pp, dp, tp) for better moe performance
-        assert (
-            self.ep_size <= self.dp_size
-        ), f"Not enough devices({self.dp_size}) for expert parallelism size({self.ep_size})."
-
-        self.moe_dp_size = self.dp_size // self.ep_size
-        self.use_ep_inside = use_ep_inside
-        if self.use_ep_inside:
-            logger.info(f"MoE Parallel use ep inside dp.", ranks=[0])
-            self.pp_axis, self.dp_axis, self.ep_axis, self.tp_axis = 0, 1, 2, 3
-            self.pg_mesh = ProcessGroupMesh(self.pp_size, self.moe_dp_size, ep_size, tp_size)
+        if dp_outside:
+            self.dp_axis, self.pp_axis, self.tp_axis, self.sp_axis = 0, 1, 2, 3
+            self.pg_mesh = ProcessGroupMesh(self.dp_size, self.pp_size, self.tp_size, self.sp_size)
+            self.moe_dp_axis, self.ep_axis = 0, 1
+            self.__moe_pg_mesh = ProcessGroupMesh(
+                self.moe_dp_size, self.ep_size, self.pp_size, self.tp_size, self.sp_size
+            )
```

Comment:
(moe_dp, pp, ep, tp, sp) may be more suitable

Reply:
Here moe_dp group and ep group are decoupled from dp_group in pg_mesh (i.e. `pg_mesh.get_group_along_axis([dp_axis]) == moe_pg_mesh.get_group_along_axis([moe_dp_axis, ep_axis])`, I guess this might not hold if we swap ep axis and pp axis?